### PR TITLE
[ClangImporter] define _GNU_SOURCE when importing glibc for Fuchsia.

### PR DESCRIPTION
**Pull request by zbowling**
Some functions and types are hidden behind feature flags.
_GNU_SOURCE exposes the most symbols and methods in glibc and musl.

(Part of https://github.com/apple/swift/pull/12955)

### CODE REVIEW

**Comment by zbowling** [comment]
@jrose-apple I pulled this out of #12955

**Comment by jrose-apple** [comment]
Thanks! I think it probably makes sense for this to just be a type-check test rather than an execution test. `%target-typecheck-verify-swift` is the thing to look for.

**Comment by zbowling** [comment]
switched to that other test type. can you ping the ci bot to run tests for me?

**Comment by CodaFi** [comment]
Serendipitously, @harlanhaskins and I just hit the issue this patch will fix.  Excellent work :sparks

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by gparker42** [comment]
The Linux failure is caused by this change. Defining `_GNU_SOURCE` changes the signature of `strerror_r()`. Apparently POSIX and GNU defined incompatible things.

The Mac failure means you need to add `os(Fuschia)` support first, I think? (It is possible to run this PR's tests with another PR applied as well, if you have a separate PR with that change and you expect them all to get merged together.)

**Comment by zbowling** [comment]
Yeah I figured the os(Fuchsia) would just emit a warning and not fail the tests on Mac until it landed (what it was doing locally). 

I wasn't seeing the strerror_r on my Ubuntu 14 machine. I'll do some investigating for a potential work around.

**Comment by zbowling** [comment]
Oh the failure was in swiftpm, not swift. I didn't test that. 🙃 I'll submit a fix to swiftpm.

**Comment by zbowling** [comment]
@gparker42 yeah I split this out #12955 which has all commits there. I was trying to land some of the changes separately so it wasn't such a big PR. Debating if that was a good idea now.

**Comment by gparker42** [comment]
Right. Might need to arrange some staging, with one or two smaller changes going in first to unblock the rest.

**Comment by zbowling** [comment]
@gparker42 i added a change to swift PM to work around that one symbol.

Unfortunately I'm gating this on #if os(Linux) which means linux systems that use musl instead of glibc won't work (like arch linux). Since, while they support most of the same _GNU_SOURCE flags, the strerror_p oddity is not preserved. Android bionic also doesn't do that but at least that is treated as another OS. 

Something to solve maybe in a future evolution proposal to ease some of libc differences.

**Comment by zbowling** [comment]
I can remove the #if os(Fuchsia) check in this commit and add it back in the Fuchsia one easy enough though.

**Comment by gparker42** [comment]
I suspect this now ought to go to swift-evolution, because `_GNU_SOURCE` has source-breaking effects other than merely providing more functions and types.

**Comment by zbowling** [comment]
Yeah, I could gate the _GNU_SOURCE define so it's only applied to specific target triples (basically not Linux) until we settle this in evolution. That would for me to get Fuchsia up and running for now. 

I ran just ran through the rest of the glibc headers to see if there are any more hidden bombs like this one and it looks pretty good. Thankfully glibc uses a macro when they define two functions we the same name under different defines.

A lot of the paths I see from a quick grep are not super concerning because most of them involve LARGEFILE64 defines that weren't setting before or even now and totally moot on 64bit anyways. A few __REDIRECT cases are where there are differences between BSD_SOURCE and GNU_SOURCE emitted functions but we weren't emitting a function for either one today because we didn't define either one (specifically linux's signal() vs sysv's signal()). 

The only breakage I can see possibly is in vfwscanf, vwscanf, vswscanf, fwscanf, wscanf, and swscanf where they accept different scan args under GNU_SOURCE  but only one is a conflicting arg (but also it's wchar and I feel bad if you are attempting to use wchar in 2017 and especially in swift for anything).

**Comment by zbowling** [comment]
Yeah, I did a diff from the generated imported module with the ide test tool. The only signature difference seems to be only the ```strerror_r()``` function. (The former lead maintainer of glibc for the past 30 years sits next to me now so I might bug him about what is up with this weird function). 

Interestingly wchar.h isn't even in the glibc.modulemap so the vfwscanf, vwscanf, vswscanf, fwscanf, wscanf, and swscanf functions are not actually an issue.

**Comment by zbowling** [comment]
@jrose-apple @gparker42 issue came up here perviously: https://github.com/apple/swift-package-manager/pull/721

**Comment by zbowling** [comment]
@gparker42 @jrose-apple I have a better idea. We could also set something like this instead which includes everything in all the POSIX/C/etc standards without the enabling the GNU specific extensions.

```
"-D_ATFILE_SOURCE=200809L",
"-D_LARGEFILE64_SOURCE=200809L",
"-D_ISOC99_SOURCE=200809L",
"-D_XOPEN_SOURCE_EXTENDED=200809L",
"-D_POSIX_SOURCE=200809L",
"-D_POSIX_C_SOURCE=200809L",
"-D_POSIX_SOURCE=700"
```

This enable a good number of the new functions without hitting the strerror_r() rewrite. We still miss out on common/good GNU extension functions like ```dup3```, ```pipe2```, and ```qsort_r``` but if we need those we can manually wrap them as needed with some c code in the platform folder and manually expose in the glibc module.

**Comment by harlanhaskins** [comment]
Does that still block `dladdr` and the like, though? IIRC it's guarded by `__USE_GNU` specifically.

**Comment by zbowling** [comment]
Yeah, dladdr will not show up without defining _GNU_SOURCE explicitly. 

Well my only other crazy idea is we just rewrite stderror_r to something else obfuscated with an APINotes hack and then make a manual wrapper in C that we manually expose to swift in the glibc module that calls the non-GNU extension version.

**Comment by jrose-apple** [comment]
I feel like this really does need more discussion at this point from more of the Linux people. This should go to swift-evolution or swift-dev. Meanwhile, I think adding the most limited set of feature macros Fuchsia needs, guarded by a target OS check, is a reasonable way to move forward.

**Comment by zbowling** [comment]
Sounds good.

**Comment by zbowling** [comment]
@jrose-apple @gparker42 I wrapped this in a triple check for Fuchsia for now. 

I kept the test since FreeBSD and Android will still expose the functions used in the test even without the feature macro defined in their libcs. For now I just marked the test to exclude linux for now until we figure out what we want to do.

**Comment by phausler** [comment]
@swift-CI please test

**Comment by zbowling** [comment]
err my force push broke the test CI it looks like

**Comment by phausler** [comment]
looks like it is just the first pass, the second one that is running now should be the right sha

**Comment by gparker42** [comment]
That CI hiccup is normal. As described in the failure log it has automatically started another CI run.

**Comment by jrose-apple** [comment]
I was wondering if you don't want to define all of `_GNU_SOURCE` on Fuchsia yet in case the conclusion is to define something more restrictive, but it's probably fine.

**Comment by zbowling** [comment]
I think I can go lower than _GNU_SOURCE on Fuchsia but I need test. Our libc is forked from musl and musl doesn't go to the extent that glibc goes with feature macros (musl strives to be compatible with glibc for that flag in a lot of cases but isn't perfect) so I need to verify something like _XOPEN_SOURCE + _POSIX_SOURCE will turn on everything we have already using in our swift code internally.

**Comment by zbowling** [comment]
I switched to "xfail: linux" instead of "unsupported: os=linux" because that seems to still fail the linux test on the CI.

**Comment by gparker42** [comment]
`OS=Linux` is not a supported test. `XFAIL: linux` should work, although that might also XFAIL the Android build too.

**Comment by zbowling** [comment]
I rebased/squashed this onto master. This should be good to test.

**Comment by gottesmm** [comment]
@swift-ci please test

**Comment by gottesmm** [comment]
@swift-ci please test

**Comment by zbowling** [comment]
merge?

File: test/stdlib/LibcFeatureFlag.swift
Comment (jrose-apple) [inline test/stdlib/LibcFeatureFlag.swift]: Since we're not actually executing the test, there's no need for this anymore…

File: test/stdlib/LibcFeatureFlag.swift
Comment (jrose-apple) [inline test/stdlib/LibcFeatureFlag.swift]: …nor do we need to actually print anything. `_ = drand48()` is sufficient.

File: lib/ClangImporter/ClangImporter.cpp
Comment (jrose-apple) [inline lib/ClangImporter/ClangImporter.cpp]: Typo: "ideally"

**Review by jrose-apple — COMMENTED** [review]
I'm good with the content at this point. A few suggestions still on the test (sorry to push this through another test cycle).

File: test/stdlib/LibcFeatureFlag.swift
Comment (zbowling) [reply test/stdlib/LibcFeatureFlag.swift]: I can kill it but I was hoping to leave it document the inconsistency/differences with glibc coverage depending on the platform you run the test on (like this test will pass on FreeBSD and Android but not Linux) until we can figure out what feature flag we want to set. Once we add a feature flag for Linux this test should start passing and we can kill the XFAIL :smile:

File: test/stdlib/LibcFeatureFlag.swift
Comment (zbowling) [reply test/stdlib/LibcFeatureFlag.swift]: yeah I wasn't it mattered since it was a %target-typecheck-verify-swift test. do we actually execute those?

File: lib/ClangImporter/ClangImporter.cpp
Comment (zbowling) [reply lib/ClangImporter/ClangImporter.cpp]: Oops

File: test/stdlib/LibcFeatureFlag.swift
Comment (jrose-apple) [reply test/stdlib/LibcFeatureFlag.swift]: :-/ "REQUIRES: executable_test" doesn't distinguish between Linux and FreeBSD, though.

File: test/stdlib/LibcFeatureFlag.swift
Comment (jrose-apple) [reply test/stdlib/LibcFeatureFlag.swift]: We don't (deliberately, to speed up testing time). Might as well not include things that aren't being tested.

File: test/stdlib/LibcFeatureFlag.swift
Comment (zbowling) [reply test/stdlib/LibcFeatureFlag.swift]: Gotcha. I'll kill the test until we figure out how to enable this flag for Linux eventually.

**Comment by zbowling** [comment]
Killed the test and reworded the comment (with the spelling fix)

**Review by jrose-apple — APPROVED** [review]

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

File: test/stdlib/LibcFeatureFlag.swift
Comment (gparker42) [reply test/stdlib/LibcFeatureFlag.swift]: As I understand it you want something like `REQUIRES: executable_test` and `UNSUPPORTED: OS=linux-gnu`.

Comment (gparker42) [reply test/stdlib/LibcFeatureFlag.swift]: (`XFAIL` always runs the test, but expects the test to fail on the specified configurations. `UNSUPPORTED` skips the test on the specified configurations.)

**Comment by jrose-apple** [comment]
Merged. Thanks, Zac.

**Pull request merged**
